src/java.base/share/classes/jdk/internal/module/ModuleInfo.java

```diff
@@ -189,6 +190,7 @@ private Attributes doRead(DataInput input) throws IOException {
 
         int minor_version = in.readUnsignedShort();
         int major_version = in.readUnsignedShort();
+        boolean isPreview = minor_version == ClassFile.PREVIEW_MINOR_VERSION;
```

Comment:
I think this will need to throw invalidModuleDescriptor if isPreview && major version != ClassFile.latestMajorVersion, otherwise the check in readModuleAttribute will be defeated by hand craft class files (the VM doesn't read module-info class files so it won't reject it, and this code is called from exposed APIs such as ModuleDescriptor.read).

Reply:
I believe the check is already being done, right on the next line:
https://github.com/openjdk/jdk/blob/821c514a132e809a14648ddbb56f2ffee85fd35a/src/java.base/share/classes/jdk/internal/module/ModuleInfo.java#L192
which leads to:
https://github.com/openjdk/jdk/blob/821c514a132e809a14648ddbb56f2ffee85fd35a/src/java.base/share/classes/jdk/internal/misc/VM.java#L188
which then leads to:
https://github.com/openjdk/jdk/blob/821c514a132e809a14648ddbb56f2ffee85fd35a/src/java.base/share/classes/jdk/internal/misc/VM.java#L174
which has this check:
https://github.com/openjdk/jdk/blob/821c514a132e809a14648ddbb56f2ffee85fd35a/src/java.base/share/classes/jdk/internal/misc/VM.java#L179

I tried with a `module-info.class` with major version 65, and minor version 0xFFFF, and it correctly threw the exception for me.

Do I miss something?

Reply:
> Do I miss something?

Okay, so this must have been modified at some point to add/use isSupportedModuleDescriptorVersion, in which case readModuleAttribute won't be called. In that case, what you have is okay.

Minor nit, can you fix up the code style in the modified expression as it has ended up with && at the start and end of lines which looks a bit messy.

Reply:
Thanks, done:
https://github.com/openjdk/jdk/pull/21431/commits/9ae10e2c43d240ba3647531c4d041589cf02875d